### PR TITLE
capture stdout to get log from solvers

### LINKED ISSUE #277: getting raw solver output
Moving e-mail conversation to an issue. Want a way to view/obtain raw output from solvers.

@pgkirsch:
...Do we want to add being able to see the number of solver iterations as a feature,...? 

For my own future reference (and for anyone else who doesn’t know/has forgotten how to do this) I’m including how I found the number of mosek iterations. Please let me know if there is a neater way that I missed. The only method I could figure out how to see the number of iterations was to add the following line before the return statement in `gpkit._mosek.expopt.printcb()`

`print msg[:-1]`

which printed the full solver log to the command line (big thank you to whoever wrote the docstring for `printcb`). Maybe the potential feature I mentioned above could be as simple as an optional argument to `solve()` that switches the statement above on?

@whoburg:
Yes, agreed we want this feature. @bqpd may have suggestions on implementation, but what you outlined sounds like a great start. Note we have to make sure it works cleanly independent of which solver is being used.

@bqpd:
Uncommenting that line seems like the right way to go! Ideally this would be controlled by the "settings" dictionary that is passed to the solver-function-constructor.

In particular, the MOSEK stream messager could have filters or flags of what to print, from everything (what you did) to nothing (the commented-out default).

### PR DESCRIPTION
**Pull request by bqpd**
Closes #277

### CODE REVIEW

**Comment by bqpd** [comment]
My thinking is that by capturing stdout, maybe we don't need those patches to the MOSEK solvers???

**Comment by bqpd** [comment]
retest this please

**Comment by bqpd** [comment]
retest this please

**Comment by whoburg** [comment]
any idea what's up with the testing? CI seems to be reporting a merge conflict

**Comment by bqpd** [comment]
test this please

**Comment by bqpd** [comment]
It's probably still behind by 1

**Comment by bqpd** [comment]
Yup.

File: gpkit/_mosek/expopt.py [hunk 1]
```diff
@@ -115,7 +115,7 @@ def printcb(void, msg):
     result : int
       0 indicates success
     """
-
+    print msg[:-1]
```
Comment (whoburg) [inline gpkit/_mosek/expopt.py]: always? what about verbosity?

**Comment by bqpd** [comment]
Argh, looks like the C code doesn't use / inherit the same stdout, so we still need those patches. :(

File: gpkit/_mosek/expopt.py (hunk 1 above)
Comment (bqpd) [inline gpkit/_mosek/expopt.py]: ;D

Comment (bqpd) [inline gpkit/_mosek/expopt.py]: Print statements during a solver call are captured and saved to the log.

**Comment by whoburg** [comment]
does this really close #277? How should a user get access to the solver output?

**Comment by whoburg** [comment]
oh I see, just look at `gp.solver_log`?

**Comment by bqpd** [comment]
Yup! Or now, just set verbosity higher.

File: gpkit/small_classes.py [hunk 2]
```diff
@@ -64,6 +64,16 @@ def __call__(self):
...
 
+class SolverLog(list):
+    "Adds a `write` method to list so it's file-like and can replace stdout."
+
+    def write(self, writ):
+        if writ != "\n":
+            if writ and writ[-1] == "\n":
+                writ = writ[:-1]
+            self.append(writ)
+
```
Comment (whoburg) [inline gpkit/small_classes.py]: haven't tested, but I think the following simpler implementation will work:

```python
def write(self, writ):
    writ = writ.strip("\n")
    if writ:
        self.append(writ)
```

Comment (bqpd) [inline gpkit/small_classes.py]: I don't want to append strings that _started_ empty, is the trick.

Comment (whoburg) [inline gpkit/small_classes.py]: Did you mean to say something else?

Mine won't append `""`.

Yours actually will append `""`...

**Comment by bqpd** [comment]
Ready for a code review!

File: gpkit/small_classes.py (hunk 2 above)
Comment (bqpd) [inline gpkit/small_classes.py]: I edited my comment after I realized that, sorry about the collision.

**Comment by bqpd** [comment]
That'd be, like 4 lines of code to __init__ list, too...

**Comment by bqpd** [comment]
Yup!

**Comment by whoburg** [comment]
retest this please

**Comment by whoburg** [comment]
this is a great fix!

**Pull request merged**